### PR TITLE
Add Faucet port tunnel ACL implementation

### PR DESCRIPTION
**Pull request by mab68**
By using a tunnel you can send packets from the applied port to another port elsewhere in the stack.

A tunnel is specified using an ACL output action. The tunnel type (only VLAN is supported), destination port, destination dp and the tunnel id (the VLAN ID) need to be specified to create the tunnel. 

All DPs will contain a dictionary of the tunnel ACLs that have been modified to represent it's position in the tunnel path (source, destination or a forwarding DP). When the DP detects an UP or DOWN update in the stack link state the DP will update the tunnel rule with the new pathways. The new tunnel rules can then be used to create a new flow-mod.

The source switch will contain a flow in the _port acl_ table that pushes the tunnel ID onto the packet using a VLAN header. The destination and forwarding switches will have flows that match the tunnel ID in the _vlan acl_ table.

_Example_: To send ipv4 icmp packets from sw1 port 1 to sw2 port 2 use the following configuration:
```
acls:
    tunnel-acl:
        - rule:
            dl_type: 0x800
            ip_proto: 1
            actions:
                output:
                    tunnel:
                        type: 'vlan'
                        tunnel_id: 200
                        dp: sw2
                        port: 2
vlans:
    vlan100:
        vid: 100
dps:
    sw1:
        dp_id: 0x1
        stack:
            priority: 1
        lldp_beacon:
            send_interval: 5
            max_per_interval: 5
        interfaces:
            1:
                native_vlan: vlan100
                acls_in: [tunnel-acl]
            2:
                receive_lldp: True
                lldp_beacon:
                    enable: True
                stack:
                    dp: sw2
                    port: 1
    sw2:
        dp_id: 0x2
        lldp_beacon:
            send_interval: 5
            max_per_interval: 5
        interfaces:
            1:
                receive_lldp: True
                lldp_beacon:
                    enable: True
                stack:
                    dp: sw1
                    port: 2
            2:
                native_vlan: vlan100
```

One thing I am not sure about is whether we want Faucet to disallow the use of certain actions and matches when using the tunnel because the tunnel ACL will be overwriting some of them. Do we want checks to prevent the user from doing that or not? 

Let me know what needs changing :)

### CODE REVIEW

**Comment by anarkiwi** [comment]
https://codecov.io/gh/faucetsdn/faucet/pull/2720/src/faucet/acl.py?before=faucet/acl.py#L324

Looks like some more coverage can be added here (codecov asserts ~10% coverage drop in acl.py).

File: faucet/dp.py [hunk 1]
```diff
@@ -821,19 +883,74 @@ def resolve_override_output_ports():
                         'override_output_port port not defined'))
                     self.ports[port_no] = port
 
-        def resolve_acl(acl_in, vid=None, port_num=None):
-            """Resolve an individual ACL."""
+        def resolve_acl(acl_in, dp=None, vid=None, port_num=None): #pylint: disable=invalid-name
```
[the remark this answers is not in the document]
Comment (mab68) [reply faucet/dp.py]: Yes, however that parameter is only used when the tunnel ACL gets created as a DP ACL and is more for distinguishing between a DP ACL and a tunnel forwarding ACL as they both do not have a port number or VID associated with resolving the ACL. The value could just be changed to a boolean or something of the sort instead if that is preferred.

File: faucet/valve.py [hunk 2]
```diff
@@ -549,6 +549,50 @@ def _update_stack_link_state(self, port, now, other_valves):
...
+                    ofmsgs.append(vlan_table.flowmod(
+                        match=vlan_table.match(vlan=tunnel_id),
+                        priority=self.dp.highest_priority,
+                        inst=[vlan_table.goto(acl_table)]
+                    ))
+                else:
+                    acl_table = self.dp.tables['port_acl']
+                    acl_allow_inst = self.pipeline.accept_to_vlan()
+                    acl_force_port_vlan_inst = self.pipeline.accept_to_l2_forwarding()
+                ofmsgs.extend(valve_acl.build_acl_ofmsgs(
```
[the remark this answers is not in the document]
Comment (KitL) [reply faucet/valve.py]: Makes sense. I am starting to think that it probably makes sense to configure the vlans in the vlan block as `reserved_internal_vlan: true` or something, kinda like how we configure ports as `output_only`. That way whatever vlan manager we create doesnt need to know anything about the tunnels, just that these vlans should only be found on stack ports and they should be pass through to the vlan acl table.

File: faucet/dp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KitL) [reply faucet/dp.py]: Would `vid is None and port_num is None:` do the trick for that?

[the remark this answers is not in the document]
Comment (mab68) [reply faucet/dp.py]: I am already using `vid is None and port_num is None` to create the forwarding tunnel acl.

**Pull request merged**